Add G425 — Auto-calibrate Backlash and Nozzle Offsets

marcio-ao opened the pull request:
This PR introduces a new feature we have been working on for our next gen printers!

This PR adds G425, a sister feature to M425, that allows for the execution of a fully automatic calibration routine. It works by touching an electrically conductive cube, bolt or washer that is mounted to the bed.

```
G425: Perform calibration with calibration cube.

B           - Perform calibration of backlash only.
T<extruder> - Perform calibration of toolhead only.
V           - Probe cube and print position, error, backlash and hotend offset.
U           - Uncertainty, how far to start probe away from the cube (mm)
no args     - Perform entire calibration sequence (backlash + position on all toolheads)
```
This feature greatly improves dual-printing by allowing the nozzle offset and backlash to be quickly calibrated to within 1/100th of a millimeter on each axis.

Get to perfect dual prints fast, with no more test prints or futzing with offsets! This feature, along with #12590, makes dual-printing a heck of a lot easier!

[![Watch the video](https://user-images.githubusercontent.com/29129419/51940420-29eb4e00-23cf-11e9-93cd-b9c6188f7a07.png)](http://devel.lulzbot.com/video/New_Marlin_Features/G425_Demo.MP4)
(click picture for video)

How it Works:
------------------

The calibration routine works as follows:

1 - It will touch the top of the cube, washer or bolt.
2 - It will then move to the side and dip the nozzle slightly below the top edge of the cube, washer or bolt.
3 - It will then tap each of the four sides to determine the center.
4 - It will repeat 1-3 for each nozzle fast to get a rough estimate of the nozzle offset and position of calibration cube, washer or bolt.
5 - It will repeat 1-3 slowly using the first nozzle to measure the backlash on X, Y, and Z.
6 - It will repeat 1-3 slowly with all nozzles to determine the measured center for each nozzle.

Once the measured center of the cube, washer or bolt is determined, the difference of this and the known actual position is used to calibrate the coordinate system and nozzle offsets to a high degree of accuracy.

Example of use and verification:
-------------------------------------------

```
> G425 ; Run automated calibration
ok
> M425 F1 S0 ; Enable full backlash compensation
> T1
> G425 V ; Re-measure cube with T1 and print report for validation

Cube Sides:␊
  Top:   2.99␊
  Left:  258.01␊
  Right: 270.02␊
  Front: -27.96␊
  Back:  -16.06␊
␊
Cube Center:␊
  X: 264.01␊
  Y: -22.01␊
  Z: -2.01␊
␊
Backlash:␊
  Left:  0.02␊
  Right: 0.02␊
  Front: 0.13␊
  Back:  0.08␊
  Top:   0.02␊
␊
Nozzle Tip Outer Dimensions:␊
  X: 2.01␊
  Y: 1.90␊
␊
T1 Positional Error:␊
  X: -0.01␊
  Y: 0.01␊
  Z: 0.01␊
␊
T1 Hotend Offset:␊
  X: 43.00␊
  Y: 0.17␊
  Z: -0.25␊
␊
ok
> T0
> G425 V ; Re-measure cube with T0 and print report for validation
␊
Cube Sides:␊
  Top:   3.02␊
  Left:  258.12␊
  Right: 269.91␊
  Front: -27.86␊
  Back:  -16.09␊
␊
Cube Center:␊
  X: 264.01␊
  Y: -21.97␊
  Z: -1.98␊
␊
Backlash:␊
  Left:  0.02␊
  Right: 0.02␊
  Front: 0.14␊
  Back:  0.08␊
  Top:   0.01␊
␊
Nozzle Tip Outer Dimensions:␊
  X: 1.79␊
  Y: 1.77␊
␊
T0 Positional Error:␊
  X: -0.01␊
  Y: -0.03␊
  Z: -0.02␊
␊
T1 Hotend Offset:␊
  X: 43.00␊
  Y: 0.17␊
  Z: -0.25␊
␊
ok
```

marcio-ao committed:
Add G425, auto-calibration

- G425 runs a fully automatic calibration routine against an electrically-conductive cube, bolt or washer mounted on the bed.

InsanityAutomation commented:
Scott is going to collapse from overwork just between our PR's this week! lol Looks amazing, ill need to test this on IDEX as well as my dualv3 thats been in a drawer for a few months! Any of the posted hex files on your repo  for the taz6 dualv3 already include this to save me a step?

marcio-ao commented:
@InsanityAutomation: No, the TAZ6 firmwares don't have this feature compiled in. However, if you use the "devel" branch you can add "`#define LULZBOT_HAS_CALIBRATION_CUBE`" to the "`#if defined(LULZBOT_Oliveoil_TAZ6)`" section in "`Conditionals_Lulzbot.h`". Then modify the section titled "AUTO-CALIBRATION (BACKLASH AND NOZZLE OFFSET)" with the proper offsets.

We haven't tested this feature with the v3, but I expect it would work.

InsanityAutomation commented:
Sounds good. Im out of town away from my machines this week, so ill plan to mess with it next week when I get back. If I can get it working nicely there, ill see what it does for idex.

marcio-ao commented:
@InsanityAutomation: I thought IDEX had two independent nozzles. Wouldn't there be no offset to speak of?

InsanityAutomation commented:
On my last calibration, I ended up with Y-.12, and the X was modified about 2.3mm from the factory value. You need to determine the spacing between endstops exactly as well. The x offset is usually the biggest one. Depending on mechanical abilities, many people struggle with setting the z offset for the second tool, hence the addition of babystepping for hotend z offset when T1 is active last year. This breaks duplicate mode though. If the hotend x calibration was used to determine difference from expected and update the hotend x offset for home position of T1 it would be highly advantageous. In duplicate mode the Y offset is irrelevant, so it starts slightly behind the other and stays that way. So its not problematic to have defined. If a Z difference is detected outside an allowable difference, idex could be configured to throw a message with M117 and the serial log that realigning the hotends vertically is highly advised if duplicate mode is intended to be used, or an error when entering duplicate mode could be thrown.

Crap.. the more I think about this the more work im making for myself!

marcio-ao commented:
@InsanityAutomation: I haven't worked with IDEX. It seems like it has its own set of challenges :)

InsanityAutomation commented:
Despite the challenges, its my preferred method. By far the most flexible dual extrusion method ive worked with. And ive worked with more methods than most, as you probably know by my varied tool change and multi extrusion commits! Duplicate mode and Mirrored duplicate mode can be quite useful for small batch parts or direct opposite parts. While youre V3 heads are nicer than the old V2's were, getting them adjusted just right was always a pain! Join the group at RAPID this year and ill steal you to my shop for a few hours to play with one lol

marcio-ao commented:
@InsanityAutomation: At the moment, I am not scheduled to go to RAPID, but we'll see :)

InsanityAutomation commented:
Just tell them you have a super secret plan to flash firmware with the lulzbot logo on every competitors machine on the showfloor and theyll bring you along for sure!

Joking aside, I would surround the setup values
```cpp
  #define CALIBRATION_MEASUREMENT_UNKNOWN           5.0                    // mm
  #define CALIBRATION_MEASUREMENT_UNCERTAIN         1.0                    // mm
  #define CALIBRATION_MEASUREMENT_CERTAIN           0.5                    // mm
```
with `ifndef` so advanced users can override them in config easily if desired.

marcio-ao commented:
> Joking aside, I would surround the setup values with ifdefs

@InsanityAutomation: Good idea! Done!

SJ-Innovation commented:
Oh thats just beautiful! What a beautifully elegant solution.

marcio-ao commented:
@thinkyhead: I only modified the main "Configuration_adv.h" file. Once you get a chance to look at it and make any necessary revisions, I can push a new commit with changes to the example config files.

thinkyhead commented:
Squashed, rebased, applied coding standards, corrected formatting, fixed some bugs:
```
git fetch origin
git checkout pr-auto-calibration-routine
git reset --hard origin/pr-auto-calibration-routine
```

thinkyhead commented:
- Messages printed to the LCD must be translatable strings. So please move all such messages to `language_en.h`.

- Some of the LCD message strings are too long, and will be cut off unless the scrolling message feature is enabled. Best practice is to keep LCD messages to within 20 characters.

  Later, it may make sense to provide a general "modal message" screen that auto-wraps and centers the given message on the screen, blocking LCD interaction until the screen is dismissed by the routine that initiated it.

  In general, it is best to keep strings very short and avoid using them altogether, if possible. Where strings must be used, try to find other code in Marlin that wants to print the same info (or similar), and see if a single function can be crafted to be used by both.

  We are at a crisis point when it comes to flash usage, so features that make heavy use of strings are discouraged.

thinkyhead commented on Marlin/src/gcode/calibrate/G425.cpp:
```diff
@@ -0,0 +1,690 @@
+        #endif
+      #endif
+    }
+};
```
I like this approach to saving values that will be restored automagically when the instance goes out of scope. We should use this technique in more places. I believe it can be generalized, so you create an instance with simply a reference to the variable being saved. The current value is stored, the new (optional) value is applied, and when it goes out of scope, it restores the saved value to the referenced variable.

I'll build this as a utility template class, play around with it, and see how it performs.

marcio-ao replied:
@thinkyhead: Yes, this is a technique I learned in a C++ book, I forget which. It's particularly useful for code which makes use exceptions, where a function could get interrupted before it would have a chance to clean up after itself. Of course, Marlin does not use exceptions, but the trick can still be useful to temporarily toggles features on and off without having to remember to restore them.

marcio-ao replied:
@thinkyhead: You modified the TemporaryBacklashCompensation class like this:

```
class TemporaryBacklashCompensation {
  public:
    TemporaryBacklashCompensation(const bool enable) {
      #if ENABLED(BACKLASH_GCODE)
        REMEMBER(backlash_correction);
        #ifdef BACKLASH_SMOOTHING_MM
          REMEMBER(backlash_smoothing_mm);
          backlash_smoothing_mm = 0;
        #endif
        backlash_correction = enable;
      #endif
    }
};
```

This is incorrect because the scope of the template that REMEMBER() instantiates is now constructor itself, so the value gets restored immediately when the constructor ends, rather than on the destructor as before. The correct way to compose the REMEMBER() macros would be like this:

```
class TemporaryBacklashCompensation {
  #if ENABLED(BACKLASH_GCODE)
  restorer<bool> a;
    #ifdef BACKLASH_SMOOTHING_MM
      restorer<float> b;
    #endif
  #endif

  public:
    TemporaryBacklashCompensation(const bool enable) : a(backlash_correction), b(backlash_smoothing_mm) {
      #if ENABLED(BACKLASH_GCODE)
        backlash_correction = enable;
        #ifdef BACKLASH_SMOOTHING_MM
          backlash_smoothing_mm = 0;
        #endif
      #endif
    }
};
```

Which is about as gnarly as it was before, so nothing much gained...

**UPDATE:** Updated with even more gnarly implementation that works, but isn't particularly satisfying.

marcio-ao replied:
Also, this macro:

```
#define REMEMBER(X) restorer<typeof(X)> X##_restorer(X)
```

Can be written as:

```
#define REMEMBER(X) restorer X##_restorer(X)
```

Since the compiler can infer the type of the template automatically.

**UPDATE:** Looks like this is only accepted in [C++ 17](https://stackoverflow.com/questions/984394/why-not-infer-template-parameter-from-constructor), so for now we can't use it that way.

marcio-ao replied:
Actually, still working on this. I thought I was compiling the code, but I wasn't. I will update the previous comments.

marcio-ao commented:
> We are at a crisis point when it comes to flash usage, so features that make heavy use of strings are discouraged.

The strings are mostly for our own benefit when using the 32-bit boards and our color LCD, where we don't have much limitations on storage. I put them in only because G425 takes so long to complete, but they are not required for the core functionality. I can remove them from G425 to save storage and add them to our own FW. I could also make most of the measurement reports optional to conserve even more PROGMEM, as they are mostly useful for debugging or troubleshooting.

> Messages printed to the LCD must be translatable strings. So please move all such messages to language_en.h.

Okay, although as I said, they aren't critical. Since space is at a premium, I'll favor removing the ones which aren't strictly needed over adding additional translations.

I'll work on tweaking this a bit.

thinkyhead commented:
Another rebase was needed. So…
```
git fetch origin
git checkout pr-auto-calibration-routine
git reset --hard origin/pr-auto-calibration-routine
```

marcio-ao commented:
@thinkyhead: I don't think the CI test is working, because there were many compilation errors I had to take care of, even though it was passing CI.

Anyhow, I made a few additional mods. I added an optional argument to your REMEMBER() macro and was able to remove the TemporaryBacklashCompensation class.

SJ-Innovation commented on Marlin/src/gcode/calibrate/G425.cpp:
```diff
@@ -0,0 +1,673 @@
+  #if ENABLED(Z_MIN_PROBE_USES_Z_MIN_ENDSTOP_PIN)
+    return (READ(Z_MIN_PIN) != Z_MIN_ENDSTOP_INVERTING);
+  #else
+    return (READ(Z_MIN_PROBE_PIN != Z_MIN_PROBE_ENDSTOP_INVERTING);
```
Issue here. Needs a closing bracket on 193. 
`return (READ(Z_MIN_PROBE_PIN != Z_MIN_PROBE_ENDSTOP_INVERTING);`
should be
`return (READ(Z_MIN_PROBE_PIN) != Z_MIN_PROBE_ENDSTOP_INVERTING);`

marcio-ao replied:
@SJ-Innovation: Good catch!

SJ-Innovation commented on Marlin/src/gcode/calibrate/G425.cpp:
```diff
@@ -0,0 +1,673 @@
+  bool saved;
+
+  public:
+    TemporaryGlobalEndstopsState(bool enable) : saved(endstops.are_endstops_enabled_globally()) {
```
Issue here on 88
`TemporaryGlobalEndstopsState(bool enable) : saved(endstops.are_endstops_enabled_globally()) {`
should be
`TemporaryGlobalEndstopsState(bool enable) : saved(endstops.global_enabled()) {`

marcio-ao replied:
@SJ-Innovation: Ah. I think it is endstops.are_endstops_enabled_globally() on my FW. I guess I need to update my code!

SJ-Innovation replied:
Its definitely `global_enabled` in the latest versions... An update might be helpful :)

SJ-Innovation commented:
Added a PR against this branch for a small hardware feature addition.

marcio-ao commented:
> Added a PR against this branch for a small hardware feature addition.

@SJ-Innovation: Oops. I wish I had seen this, as I just pushed modifications to all the Config files.

marcio-ao commented:
@thinkyhead: @InsanityAutomation: Squashed all commits since the history was getting hairy. I also went through one more time and renamed a lot of things for clarity and/or conciseness.

thinkyhead committed:
Cosmetic tweaks

thinkyhead commented:
Did another squash/rebase.

thinkyhead committed:
Allow REMEMBER to refer to complex variables

marcio-ao commented:
@thinkyhead: It looks like at some point during all this refactoring, the macros that were saving the backlash state got substituted with macros that were saving the endstop state. I think I may have done this by mistake during my last clean up. Should be back to what it needs to be now.

marcio-ao committed:
Fixed places where backlash state should be saved.

- The places where the endstop state was saved rather than the backlash state.

thinkyhead commented:
> macros that were saving the endstop state

A squirrel appeared outside my window!

thinkyhead commented:
I like the design pattern of temporary states. I'm thinking a lot about design patterns and bringing in more support classes to handle the mundane things — a thin string wrapper class, etc. And in the high-level view, I want to employ design patterns more, with the UI following an MVC pattern, the rest of Marlin taking on the model role. Marlin is also a state machine (well, many of them), so there are principles from that milieu we can apply. SRAM is scarce, but maybe we can spare some for an event dispatcher, and maybe it will save SRAM on the back end.

More coffee will be needed….

thinkyhead commented:
I just realized, a temporary state class is a perfect solution to the multiple serial port output code ugliness we now have in `configuration_store.cpp` and elsewhere. Currently we have to pass a `port` index to the serial print functions. But using the temporary state, we can set the current (or preferred) serial output port to use inside a scope, and all serial output functions will use that global `port` index within that scope, so long as this method is followed consistently when the serial output needs to be different from the default.

I'm going to put that together and see how it flies!

oysteinkrog commented:
@marcio-ao I want to add this to my own printer, but I'm wondering; does the type of calibration object make a difference? You mention bolt and washer as possible objects... is that less precise than using a cube?

lixxbox commented:
Looks great! 
Could this be used to determine the Z-offset between a nozzle and a sensor (e.g. a BLtouch) in a single nozzle system?

SJ-Innovation commented:
@lixxbox Could be, i dont think it has the ability to natively, it shouldnt be too difficult to write a routine that determines this though. Ill look into it.

marcio-ao commented:
> I want to add this to my own printer, but I'm wondering; does the type of calibration object make a difference? You mention bolt and washer as possible objects... is that less precise than using a cube?

@oysteinkrog:  The biggest cause of failures will be if the nozzle slides up on the object rather than making contact. So nice sharp straight edges are best. This is why a cube is the ideal surface. If you had a flat washer with good sharp edges, this work probably also work, but due to the curvature of the sides, you would have to hit it precisely at the mid-line. A bolt might work, but if it had chamfered egdes, it might get pushed down as the nozzle touched it. So, to summarize, a cube with nice sharp edges is the best shape, but other things might work, possibly with less accuracy and/or repeatability.

marcio-ao commented:
> Could this be used to determine the Z-offset between a nozzle and a sensor (e.g. a BLtouch) in a single nozzle system?

@lixxbox: Assuming both the sensor and the nozzle can make electrical contact with the object, yes. But the code currently only modifies the backlash setting and nozzle_offset. It would not be difficult to modify so it adjusts other offsets, however. There is basically one function which probes the object on all sides and it stores the located center of the object in a data structure. If you repeat this twice, once with the nozzle and once with the Z-probe, the difference between the measured center points is how much the offset would need to be adjusted. There are functions which does this for the nozzle_offset, but it could be duplicated and tweaked to adjust other offsets.

oysteinkrog commented:
Ok, that leaves me wondering where I could get such a cube :P 
I'm guessing that manufacturing it with hand tools would probably be quite difficult.
Does anyone have any good ideas?
Edit: maybe a nice square nut could work.
Also, a small cheap (square?) gage block would be an option, just have to find one..

marcio-ao commented:
> Ok, that leaves me wondering where I could get such a cube :P
> I'm guessing that manufacturing it with hand tools would probably be quite difficult.
> Does anyone have any good ideas?

@oysteinkrog:

You mean you don't have a CNC mill? That's how we did it for our prototypes ;)

You could probably make a "cube" by getting a long piece of aluminum stock with a square or rectangular profile, cut a bit off the end to make a small "cube" (the dimensions do not need to be the same on all sides, so a "rectangular cube" would work), then drill a hole on a side so you can mount it on the bed somewhere.

If your nozzle is non-magnetic, you could probably also use a small neodymium magnet -- not for the magnetic properties, but because they are easily obtained as small cubes.

marcio-ao commented:
> > macros that were saving the endstop state
> A squirrel appeared outside my window!

@thinkyhead: I don't even think it was you. We have squirrels too...
